### ISSUE TITLE
[Bug] Car "backfire" effect missing when using controllers

### ISSUE THREAD

**Issue by lance8765** [emoose/OutRun2006Tweaks#10]
I apologize in advance since the only evidence of this bug is from the FXT forums, but a real solution was never implemented as it was reported probably years after the author stopped working on it.

Basically, the cars are missing the backfire effect when upshitting using controllers. It's not missing when playing with a keyboard.

This issue happens on both vanilla and modded versions of the game.

Apparently, it's a bug related to the use of Direct Input devices. Using external programs such as XInput Plus fixes the issue. Or, as the other person said, by assigning accelerate/brake to face buttons (not triggers)

I wonder if there's a way to fix this issue without having to use external programs and/or without having to reassign the controls.

**Comment by lance8765** [comment]
I'm sorry, but i can't test it at the moment.

Looks like i'm missing a .DLL file named **api-ms-win-core-winrt-string-l1-1-0.dll** (and probably many others)

The problem is that there are many topics about that issue but many solutions and none of them work, they tell to install Windows update that i have already installed, too.

**Comment by DonelBueno** [comment]
It worked fine here, but I also have an Xbox Series controller.

**Comment by emoose** [comment]
@lance8765 you can probably remove x1nput fine if you're just using x360 controller, in future this might have something like it built in but it'll be optional, do you know if the backfire fix is working with it removed?

**Comment by lance8765** [comment]
> @lance8765 you can probably remove x1nput fine if you're just using x360 controller, in future this might have something like it built in but it'll be optional, do you know if the backfire fix is working with it removed?

I'm sorry, but the game cannot start without the X1nput files installed. It asks for the DLL.

**Comment by emoose** [comment]
Ah seems the new vibration code made it depend on xinput1_4.dll, looks like that's only available in win8+ for some reason.

Seems I can change it to use xinput9_1_0.dll which win7 should include apparently, maybe see if this build works for you: 
[OutRun2006Tweaks-0.2.5-FixFullPedalChecks-v3.zip](https://github.com/user-attachments/files/15912778/OutRun2006Tweaks-0.2.5-FixFullPedalChecks-v3.zip)

**Comment by lance8765** [comment]
Ok... this one works! Thank you.

I can confirm that both the backfire and vibration work on my 360 controller.

However... the controls are completely messed up and hard to remap, but that's an old story...

**Comment by emoose** [comment]
Hm it might be something up with the win7 x360 directinput driver, apparently that might map triggers different to how win10+ maps them.

Guess win7 way might not be compatible, weird though since you'd think a game released around the same time would have supported it fine...

Xidi might be able to change the mapping, if you rename the outrun2006tweaks dinput8.dll to winmm.dll, you could extract xidi as dinput8.dll and use the xidi.ini from https://github.com/emoose/OutRun2006Tweaks/issues/7#issuecomment-2166208309

Not really sure if xidi has win7 support or not though, but worth a try. Will try seeing if there's anything we could do on the game side for it as well.

**Comment by lance8765** [comment]
Well the game sees "Xidi Virtual Controller" instead of Xbox 360 controller, but i don't see any real benefit. Both triggers are still seen as "Z Axis 1".

**Comment by emoose** [comment]
Hm, maybe one of the other mappers could help, you can change the `XInputSharedTriggers` in xidi.ini to `ExtendedGamepad` or `XInputNative`, hopefully one might work better
There's a couple others at https://github.com/samuelgr/Xidi#built-in-mappers too but those might disable analogs

**Comment by lance8765** [comment]
I've only tried ExtendedGamepad for now. It works! Right trigger is "X-Axis 2" while the left one is "Y-Axis 2". Thanks for the tip. 

However i've just discovered a weird thing... Xbox and PS2 have a different button layout. The Xbox controller has the triggers but no LB/RB! But on PS2 you could use both L/R buttons to gear shift up or down (and IIRC, X to accelerate and Square to brake, among other things)

https://archive.org/details/xboxmanual_Outrun_2006_Coast_2_Coast/page/n1/mode/2up
https://segaretro.org/index.php?title=File:OutRun2006_PS2_US_Manual.pdf&page=8

In other words it's impossible to replicate the very same button layout as of console, on PC.

Also, in OutRun2SP mode, if you want a different color for your car, if you use the thumbstick to select the colors, the vertical axis is inverted (i.e. you press up, it goes down). It doesn't happen if you use the directional pad though. This also applies to XInputSharedTriggers (or no Xidi file at all)

**Comment by Dekaid0011** [comment]
Another issue with triggers:

When using an xinput controller with triggers as gas and brakes holding both at the same time cancels both inputs out, the game treats it like neither are pressed. The expected result has the brake lights lighting up which is currently not the case atleast using xinput triggers, using buttons instead allows for both gas and brakes at the same time. I tested this with a wired xbox 360 controller and a ds3 using dshidmini in xinput mode.

On xbox c2c atleast this worked with triggers, I tried xidi with the ini set as `ExtendedGamepad` to see if dinput would work and it did, but gas and brake axis were also halved in travel, they only started registering midway through. `XInputSharedTriggers` had the same result as normal XInput and `XInputNative` had the same result as `ExtendedGamepad` until it crashed the game when I tried to reconfigure.

**Comment by emoose** [comment]
Was able to use Xidi to fix the shared triggers:
- rename tweaks dinput8.dll to winmm.dll
- extract xidi `Win32\dinput8.dll` file to game folder
- create xidi.ini with the following

```
[Mapper]
Type=OutRunMapper

[CustomMapper:OutRunMapper]
Template=XInputNative

[Workarounds]
ActiveVirtualControllerMask=1
```

- open OutRun2006Tweaks.ini, set ControllerHotPlug to false
- run game, start race
- at this point the accelerator pedal should be stuck, pressing LT will decelerate and apply brake at the same time
- pause game > controls > configuration
- the first page should be Xidi Virtual Controller, and it should list Accelerate Axis/Brake Axis/etc
- move down to Accelerate axis, press A, then press/release RT a few times until it mentions "will be assigned to: Axis 2 Z", then press A
- when it asks for minimum/rest position, press RT and release it a few times, until the bars are empty, then press A
- when it asks for maximum/full, press RT and release, then hold RT until bars fill, then press A
- do the same for Brake Axis
- once Accelerate & Brake are set, press B to exit out, then see if triggers work ingame
